cpp/src/arrow/compute/kernels/scalar_cast_string.cc

```diff
@@ -92,12 +93,74 @@ struct Utf8Validator {
 };
 
 template <typename I, typename O>
-struct BinaryToStringSameWidthCastFunctor {
+constexpr bool has_smaller_width() {
+  return sizeof(typename I::offset_type) < sizeof(typename O::offset_type);
+}
+
+template <typename I, typename O>
+constexpr bool has_same_width() {
+  return sizeof(typename I::offset_type) == sizeof(typename O::offset_type);
+}
+
+// Cast same-width offsets (no-op)
+template <typename I, typename O>
+void CastBinaryToBinaryOffsets(enable_if_t<has_same_width<I, O>(), KernelContext*> ctx,
+                               const ArrayData& input, ArrayData* output) {}
+
+// Upcast offsets
+template <typename I, typename O>
+void CastBinaryToBinaryOffsets(enable_if_t<has_smaller_width<I, O>(), KernelContext*> ctx,
+                               const ArrayData& input, ArrayData* output) {
```

Comment:
```suggestion
template <typename I, typename O>
void CastBinaryToBinaryOffsets(enable_if_t<has_smaller_width<I, O>() || has_smaller_width<O, I>(), KernelContext*> ctx,
                               const ArrayData& input, ArrayData* output) {
  if (has_smaller_width<I, O>()) {
    //
  } else {
    //
  }
```

Reply:
Lack of `constexpr if` risks producing compile errors.

Reply:
If constexpr can be emulated using `std::conditional<has_smaller_width<I, O>(), NarrowerImpl, WiderImpl>::template Exec<I, O>(ctx, input, output)`

Reply:
Horror.